engine/table/src/main/java/io/deephaven/engine/table/impl/SliceLikeOperation.java

```diff
@@ -92,31 +95,40 @@ protected long getLastPositionExclusive() {
-        resultTable = parent.getSubTable(resultRowSet);
-        if (isFlat) {
-            resultTable.setFlat();
-        }
-
-        if (operation.equals("headPct")) {
-            // headPct has a floating tail, so we can only propagate if append-only
-            if (parent.isAppendOnly()) {
-                resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
-                resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+        if (!parent.isBlink()) {
+            final TrackingRowSet parentRowSet = parent.getRowSet();
+            final TrackingRowSet resultRowSet =
+                    computeSliceIndex(usePrev ? parentRowSet.prev() : parentRowSet).toTracking();
+            // result table must be a sub-table, so we can pass ModifiedColumnSet to listeners when possible
+            resultTable = parent.getSubTable(resultRowSet);
+            if (isFlat) {
+                resultTable.setFlat();
             }
-        } else if (!operation.equals("tailPct") && getFirstPositionInclusive() >= 0) {
-            // tailPct has a floating head, so we can't propagate either property
-            // otherwise, if the first row is fixed (not negative), then we can propagate add-only/append-only
-            if (parent.isAddOnly()) {
-                resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+
+            if (operation.equals("headPct")) {
+                // headPct has a floating tail, so we can only propagate if append-only
+                if (parent.isAppendOnly()) {
+                    resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+                    resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+                }
+            } else if (!operation.equals("tailPct") && getFirstPositionInclusive() >= 0) {
+                // tailPct has a floating head, so we can't propagate either property
+                // otherwise, if the first row is fixed (not negative), then we can propagate add-only/append-only
+                if (parent.isAddOnly()) {
+                    resultTable.setAttribute(Table.ADD_ONLY_TABLE_ATTRIBUTE, true);
+                }
+                if (parent.isAppendOnly()) {
+                    resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+                }
             }
-            if (parent.isAppendOnly()) {
-                resultTable.setAttribute(Table.APPEND_ONLY_TABLE_ATTRIBUTE, true);
+        } else {
+            if (operation.equals("head")) {
+                long size = getLastPositionExclusive();
+                resultTable = (QueryTable) BlinkTableTools.blinkToAppendOnly(parent, size);
+            } else if (operation.equals("tail")) {
+                long size = -1 * getFirstPositionInclusive();
+                resultTable = (QueryTable) RingTableTools.of(parent, (int) size); // TODO Do I need to do something for
+                // this type cast from long to int?
```

Comment:
I am not sure how to handle this type case without making a big change in RingTableTools.

Reply:
You should do a conversion that will cause exception on overflow.  `Math.toIntExact()` would be a reasonable choice.

Reply:
Exactly what I was looking for, thanks.